wire-schema/src/main/java/com/squareup/wire/schema/Options.kt

```diff
@@ -55,26 +63,30 @@ class Options(
   fun optionMatches(namePattern: String, valuePattern: String): Boolean {
     val nameMatcher = Pattern.compile(namePattern).matcher("")
     val valueMatcher = Pattern.compile(valuePattern).matcher("")
-    return map!!.any { entry ->
-      nameMatcher.reset(entry.key.member).matches() &&
+
+    return entries!!.any { entry ->
+      nameMatcher.reset(entry.protoMember.member).matches() &&
           valueMatcher.reset(entry.value.toString()).matches()
     }
   }
 
   fun link(linker: Linker) {
-    var map = ImmutableMap.of<ProtoMember, Any?>()
-    for (option in elements) {
-      val canonicalOption = canonicalizeOption(linker, optionType, option) ?: continue
-      map = union(linker, map, canonicalOption)
+    var entries: List<LinkedOptionEntry> = emptyList()
+
+    for (option in optionElements) {
+      val canonicalOption: List<LinkedOptionEntry> = canonicalizeOption(linker, optionType, option)
+              ?: continue
+
+      entries = union(linker, entries, canonicalOption)
```

Comment:
For what it’s worth, this is `O(N*N)` which is bad if anyone has lots of options

Reply:
We can look to optimise later. It's already N*N I believe.